stbt-run: Allow running tests by test name

drothlis opened the pull request:
You can now run a test with:

     stbt run test.py::my_test_function_name

This mimics py.test's command line interface, except that `stbt run test.py` won't run all `test_*` functions in the file.

This brings `stbt` to parity with the stb-tester ᴏɴᴇ appliance.

TODO:

* [x] Decide whether we should scrap this and figure out how to get py.test to create `stbt batch`-style reports, or whether we need to support this in `stbt run` as an interim measure.
* [x] Selftest that `stbt batch` works with this style of test name.
* [x] Release notes.
* [x] Review release notes wording.
* [x] Update stbt record to create a script in this style.
* [x] Update `docs/getting-started.md` etc.
* ~~Update tab-completion script for `stbt run` and `stbt batch run`.~~

wmanley commented:
If we're going to recommend this approach above one test-case per file we really need to update all our documentation (and `stbt record`) to reflect that.  What do you think?

wmanley commented:
Also, I'm not that happy with the release-notes wording.

drothlis commented:
> If we're going to recommend this approach above one test-case per file we really need to update all our documentation (and `stbt record`) to reflect that. What do you think?

Agreed. I've added those to the todo list.

drothlis commented:
I rebased onto master, squashed Will's fixups, added my own fixups to
the release notes (unsquashed for easier review), taught `stbt record`
to produce python files in the new format, and updated
`docs/getting-started.md`.

The `stbt record` change is a slight compatibility break to people's
manual workflows (if they want to keep the old `stbt run` command line
they'll have to edit the script generated by `stbt record` to remove the
function), but I can't imagine there will be any automated process
affected by this change.

drothlis commented:
I think the remaining todo item in this pull request (Decide whether we should scrap this and figure out how to use pytest instead) can be decided in favour of merging this, as it allows us to get people using this format (which is forward-compatible with pytest) now. @wmanley if you're happy with this I'd like to merge it (after squashing the fixups of course).

wmanley commented:
> I think the remaining todo item in this pull request (Decide whether we should scrap this and figure out how to use pytest instead) can be decided in favour of merging this, as it allows us to get people using this format (which is forward-compatible with pytest) now.

Agreed

> @wmanley if you're happy with this I'd like to merge it (after squashing the fixups of course).

+1

drothlis commented:
I'm unlikely to finish the bash tab-completion script today so I'm going to merge this now.

drothlis commented:
I'll do the tab-completion in #253.
